### PR TITLE
Added Groups API documentation

### PR DESCRIPTION
**Pull request by aleksandrychev**
Ticket: ENT-10196

### CODE REVIEW

**Comment by aleksandrychev** [comment]
@cf-bottom jenkins

**Comment by aleksandrychev** [comment]
@nickanderson please review

**Comment by aleksandrychev** [comment]
[Build Status](https://ci.cfengine.com/job/fr-tests/6095/)

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/fast-build-and-deploy-docs-master/364/)

**Jenkins:** https://ci.cfengine.com/job/fast-build-and-deploy-docs-master/364/

**Documentation:** http://buildcache.cfengine.com/packages/build-documentation-pr/jenkins-fast-build-and-deploy-docs-master-364/output/_site/

File: api/enterprise-api-ref/shared-groups.markdown [hunk 1]
```diff
@@ -0,0 +1,663 @@
...
+
+* **filter** *(json object)*  Group filter object. Includes inventory filter and classes filters
+  * **filter** *(json object)*  Optional parameter.
+    Inventory filter data. You can use array values for multiple filter, the logic will be AND. Format is
+  * **hostContextInclude** *(array)* Optional parameter.
+    Includes only results that concern hosts which have all specified CFEngine contexts (class) set.
+  * **hostContextExclude** *(array)*  Optional parameter.
+    Excludes results that concern hosts which have specified CFEngine context (class) set. Hosts that have at least one of the specified contexts set will be excluded from the results.
+
+```
```
Comment (nickanderson) [inline api/enterprise-api-ref/shared-groups.markdown:34]:
```suggestion
```json
```

File: api/enterprise-api-ref/shared-groups.markdown [hunk 2]
```diff
@@ -0,0 +1,663 @@
...
+                "aix"
+            ]
+        }
+  }'
+```
+
+
+**Example response:**
+
+```
```
Comment (nickanderson) [inline api/enterprise-api-ref/shared-groups.markdown:93]:
```suggestion
```json
```

File: api/enterprise-api-ref/shared-groups.markdown [hunk 3]
```diff
@@ -0,0 +1,663 @@
...
+
+* **filter** *(json object)*  Group filter object. Includes inventory filter and classes filters
+  * **filter** *(json object)*  Optional parameter.
+    Inventory filter data. You can use array values for multiple filter, the logic will be AND. Format is
+  * **hostContextInclude** *(array)* Optional parameter.
+    Includes only results that concern hosts which have all specified CFEngine contexts (class) set.
+  * **hostContextExclude** *(array)*  Optional parameter.
+    Excludes results that concern hosts which have specified CFEngine context (class) set. Hosts that have at least one of the specified contexts set will be excluded from the results.
+
+```
```
Comment (nickanderson) [inline api/enterprise-api-ref/shared-groups.markdown:126]:
```suggestion
```json
```

File: api/enterprise-api-ref/shared-groups.markdown [hunk 4]
```diff
@@ -0,0 +1,663 @@
...
+                "aix"
+            ]
+        }
+  }'
+```
+
+
+**Example response:**
+
+```
```
Comment (nickanderson) [inline api/enterprise-api-ref/shared-groups.markdown:187]:
```suggestion
```json
```

File: api/enterprise-api-ref/shared-groups.markdown [hunk 5]
```diff
@@ -0,0 +1,663 @@
...
+curl -k --user <username>:<password> \
+  -X GET \
+  https://hub.cfengine.com/api/host-groups/shared/4 \
+  -H 'content-type: application/json'
+```
+
+**Example response:**
+
+```
```
Comment (nickanderson) [inline api/enterprise-api-ref/shared-groups.markdown:213]:
```suggestion
```json
```

File: api/enterprise-api-ref/shared-groups.markdown [hunk 6]
```diff
@@ -0,0 +1,663 @@
...
+curl -k --user <username>:<password> \
+  -X GET \
+  https://hub.cfengine.com/api/host-groups/shared\
+  -H 'content-type: application/json'
+```
+
+**Example response:**
+
+```
```
Comment (nickanderson) [inline api/enterprise-api-ref/shared-groups.markdown:275]:
```suggestion
```json
```

File: api/enterprise-api-ref/shared-groups.markdown [hunk 7]
```diff
@@ -0,0 +1,663 @@
...
+curl -k --user <username>:<password> \
+  -X POST \
+  https://hub.cfengine.com/api/host-groups/shared/5/makePersonal
+```
+
+**Example response:**
+
+API returns new ID of the personal group.
+
+```
```
Comment (nickanderson) [inline api/enterprise-api-ref/shared-groups.markdown:341]:
```suggestion
```json
```

**Review by nickanderson — CHANGES_REQUESTED** [review]

File: api/enterprise-api-ref/personal-groups.markdown [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+        "description":"Host name",
+        "filter":{
+            "filter": {
+                "Architecture": {
+                    "matches": "86"
+                }
+            },
+            "hostsFilter": {
+                "includes": {
+                    "type": "hostkey",
```
[the remark this answers is not in the document]
Comment (aleksandrychev) [reply api/enterprise-api-ref/personal-groups.markdown:81]: only hostkey for now

File: api/enterprise-api-ref/personal-groups.markdown [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+| is_reported     |
+| is_not_reported |
+
+
+**Example request:**
+
+```
+curl -k --user <username>:<password> \
+  -X PATCH \
+  https://hub.cfengine.com/api/host-groups/personal/ \
```
[the remark this answers is not in the document]
Comment (aleksandrychev) [reply api/enterprise-api-ref/personal-groups.markdown:165]: fixed

File: api/enterprise-api-ref/personal-groups.markdown [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+        "name":"AIX hosts",
+        "description":"Host name",
+        "filter":{
+            "filter": {
+                "Architecture": {
+                    "matches": "86"
+                }
+            },
+            "hostsFilter": {
+                "includes": {
```
[the remark this answers is not in the document]
Comment (aleksandrychev) [reply api/enterprise-api-ref/personal-groups.markdown:80]: you need to wrap all your filters into `filter` property. but hostsFitlers adds it additionally to existing inv or class filters

File: api/enterprise-api-ref/personal-groups.markdown [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+
+```
+{"id":"4"}
+```
+
+## Update group
+
+**URI:** https://hub.cfengine.com/api/host-groups/personal/:id
+
+**Method:** PATCH
```
[the remark this answers is not in the document]
Comment (aleksandrychev) [reply api/enterprise-api-ref/personal-groups.markdown:105]: no minimal, just wrap hostsFilter into filter

**Review by olehermanse — CHANGES_REQUESTED** [review]

File: api/enterprise-api-ref/personal-groups.markdown [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+  -d '{
+        "name":"AIX hosts",
+        "description":"Host name",
+        "filter":{
+            "filter": {
+                "Architecture": {
+                    "matches": "86"
+                }
+            },
+            "hostsFilter": {
```
[the remark this answers is not in the document]
Comment (aleksandrychev) [reply api/enterprise-api-ref/personal-groups.markdown:107]: added

File: api/enterprise-api-ref/personal-groups.markdown [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+        Object that specifies hosts to be excluded.
+          * **type** (string)
+            Now the only supported type is `hostkey`.
+          * **data** *(array)*
+            An array of hosts keys.
+
+```json
+{ 
+  "filter": {
+      "filter":{
```
Comment (nickanderson) [inline api/enterprise-api-ref/personal-groups.markdown:172]: This filter.filter thing was a bit confusing to me. For whatever reason I wasn't seeing both, making me think that hostsFilter and hostcontextinclude/exclude were outside of filter.

File: api/enterprise-api-ref/shared-groups.markdown [hunk 14]
```diff
@@ -0,0 +1,720 @@
...
+        "timestamp": 1619116399,
+        "total": "1",
+        "cmdb_epoch": "13"
+    }
+}
+```
+
+## Get group's specific configuration
+
+**URI:** https://hub.cfengine.com/api/host-groups/shared/:id/cmdb/:type/:name/
```
Comment (nickanderson) [inline api/enterprise-api-ref/shared-groups.markdown:477]: Is it expected for this trailing slash to be strictly required?

I was struggling with the API for a while until I finally realized I was missing the trailing `/` things worked as soon as I added it.

File: api/enterprise-api-ref/shared-groups.markdown [hunk 15]
```diff
@@ -0,0 +1,720 @@
...
+
+**Example response:**
+
+```
+HTTP 200 Ok
+```
+
+
+
+## Delete host's configurations
```
Comment (nickanderson) [inline api/enterprise-api-ref/shared-groups.markdown:665]: Delete host's configurations? Or delete /group's/ configurations?

**Pull request merged**